### ISSUE TITLE
Some floating-point stuff broke during the z3 upgrade

### ISSUE THREAD

**Issue by rhelmot** [angr/claripy#86]
I just upgraded the z3 distribution to the latest version and, as per usual, it seems to have broken a few things. The following code worked before the upgrade and no longer does:

```python
import claripy
s = claripy.Solver()
a = claripy.FPS('a', claripy.fp.FSORT_FLOAT)
s.eval(a.raw_to_bv(), 1)
```

Something along these lines is currently causing our CI to fail.

I would appreciate it if someone could take this issue on.

**Comment by Mic92** [comment]
Are you sure you can put a string into FPV? Because this will crash in this line: https://github.com/angr/claripy/blob/master/claripy/backends/backend_z3.py#L270

```
claripy.FPV('a', claripy.fp.FSORT_FLOAT)
```

**Comment by rhelmot** [comment]
Sorry, that was a typo - the actual testcase is with FPS, not FPV. I've edited the issue. The intent is to just do the simplest possible thing that will push something into and out of z3.

**Comment by Mic92** [comment]
I suppose that was meant:

```console
$ nosetests --nologcapture tests/test_callable.py:test_manyfloatsum_symbolic
E
.
======================================================================
ERROR: test_callable.test_manyfloatsum_symbolic('i386',)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/joerg/git/hase/.direnv/python-2.7.15/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/joerg/git/hase/angr/tests/test_callable.py", line 94, in run_manyfloatsum_symbolic
    args_conc = s.batch_eval(args, 1)[0]
  File "/home/joerg/git/hase/claripy/claripy/frontend_mixins/concrete_handler_mixin.py", line 18, in batch_eval
    super(ConcreteHandlerMixin, self).batch_eval(symbolic_exprs, n, **kwargs)
  File "/home/joerg/git/hase/claripy/claripy/frontend_mixins/constraint_filter_mixin.py", line 44, in batch_eval
    return super(ConstraintFilterMixin, self).batch_eval(exprs, n, extra_constraints=ec, **kwargs)
  File "/home/joerg/git/hase/claripy/claripy/frontend_mixins/sat_cache_mixin.py", line 70, in batch_eval
    extra_constraints=extra_constraints, **kwargs
  File "/home/joerg/git/hase/claripy/claripy/frontend_mixins/model_cache_mixin.py", line 255, in batch_eval
    asts, remaining, extra_constraints=constraints, **kwargs
  File "/home/joerg/git/hase/claripy/claripy/frontend_mixins/simplify_helper_mixin.py", line 18, in batch_eval
    return super(SimplifyHelperMixin, self).batch_eval(e, n, *args, **kwargs)
  File "/home/joerg/git/hase/claripy/claripy/frontends/full_frontend.py", line 108, in batch_eval
    if not self.satisfiable(extra_constraints=extra_constraints):
  File "/home/joerg/git/hase/claripy/claripy/frontend_mixins/constraint_filter_mixin.py", line 34, in satisfiable
    return super(ConstraintFilterMixin, self).satisfiable(extra_constraints=ec, **kwargs)
  File "/home/joerg/git/hase/claripy/claripy/frontend_mixins/sat_cache_mixin.py", line 45, in satisfiable
    extra_constraints=extra_constraints, **kwargs
  File "/home/joerg/git/hase/claripy/claripy/frontend_mixins/model_cache_mixin.py", line 235, in satisfiable
    return super(ModelCacheMixin, self).satisfiable(extra_constraints=extra_constraints, **kwargs)
  File "/home/joerg/git/hase/claripy/claripy/frontends/full_frontend.py", line 90, in satisfiable
    solver=self._get_solver(), model_callback=self._model_hook
  File "/home/joerg/git/hase/claripy/claripy/backends/__init__.py", line 567, in satisfiable
    return self._satisfiable(extra_constraints=self.convert_list(extra_constraints), solver=solver, model_callback=model_callback)
  File "/home/joerg/git/hase/claripy/claripy/backends/backend_z3.py", line 592, in _satisfiable
    model_callback(self._generic_model(solver.model()))
  File "/home/joerg/git/hase/claripy/claripy/backends/backend_z3.py", line 570, in _generic_model
    m = m_f()
  File "/home/joerg/git/hase/angr-z3/src/api/python/z3/z3.py", line 751, in __call__
    _z3_assert(num == self.arity(), "Incorrect number of arguments to %s" % self)
  File "/home/joerg/git/hase/angr-z3/src/api/python/z3/z3.py", line 91, in _z3_assert
    raise Z3Exception(msg)
Z3Exception: Incorrect number of arguments to fp.to_ieee_bv
----------------------------------------------------------------------
```

**Comment by rhelmot** [comment]
Yeah, that looks like a similar problem. Z3 keeps changing the API for its floating point functions and I'm sick of adapting to it at this point...

**Comment by Mic92** [comment]
It might be this commit: https://github.com/Z3Prover/z3/commit/31cfca0444d2df3f231e0a8b92e80a8fa531f7c1

**Comment by saullocarvalho** [comment]
Is this still a issue? How can I help?

**Comment by Phat3** [comment]
I just created a PR (#145 ) that should fix all the problem that we had with the newer version of Z3.

The problem that we had was when we wanted to perform a conversion from a symbolic FP to a BV. Under the hood, angr calls the function FPToIEEE with the FP as a parameter. In the old version of z3, there was a concept of the function FPToIEE_unspecified which was special cased to 0 while in the new version this unspecified version of the function is no longer present.

Because of this, we ended up having two different models for the same constraint in the two different versions of Z3 (old version: `[Sym_FP = NaN, fp.to_ieee_bv_unspecified = 0]`, new version: `[Sym_FP = NaN]`).
